generator/validate/gen.go

```diff
@@ -0,0 +1,178 @@
+package validate
+
+import (
+	"fmt"
+	"go/ast"
+	"go/types"
+	"strings"
+
+	"github.com/devfile/api/generator/genutils"
+	crdmarkers "sigs.k8s.io/controller-tools/pkg/crd/markers"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/loader"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+//go:generate go run sigs.k8s.io/controller-tools/cmd/helpgen generate:headerFile=../header.go.txt,year=2020 paths=.
+
+// +controllertools:marker:generateHelp
+
+// Generator validates the consistency of the API GO code.
+//
+// Validity checks are related to unions, patchStrategy, and optional fields.
+type Generator struct{}
+
+// RegisterMarkers registers the markers of the Generator
+func (Generator) RegisterMarkers(into *markers.Registry) error {
+	err := genutils.RegisterUnionMarkers(into)
+	if err != nil {
+		return err
+	}
+	return crdmarkers.Register(into)
+}
+
+// Generate validates the source code
+func (g Generator) Generate(ctx *genall.GenerationContext) error {
+	for _, root := range ctx.Roots {
+		ctx.Checker.Check(root, func(node ast.Node) bool {
+			// ignore interfaces
+			_, isIface := node.(*ast.InterfaceType)
+			return !isIface
+		})
+
+		root.NeedTypesInfo()
+
+		packageTypes := map[string]*markers.TypeInfo{}
+		if err := markers.EachType(ctx.Collector, root, func(info *markers.TypeInfo) {
+			packageTypes[info.Name] = info
+		}); err != nil {
+			root.AddError(err)
+			return nil
+		}
+
+		for _, typeToCheck := range packageTypes {
+			checkUnion(typeToCheck, root, packageTypes)
+		}
+
+	}
+
+	return nil
+}
+
+func checkUnion(typeToCheck *markers.TypeInfo, root *loader.Package, packageTypes map[string]*markers.TypeInfo) {
+	if typeToCheck.Markers.Get(genutils.UnionMarker.Name) == nil {
+		return
+	}
+
+	unionMembers := []string{}
+	var unionDiscriminatorField *markers.FieldInfo
+	for index, unionField := range typeToCheck.Fields {
+		if unionField.Markers.Get(genutils.UnionDiscriminatorMarker.Name) != nil {
+			if unionDiscriminatorField != nil {
+				root.AddError(loader.ErrFromNode(fmt.Errorf(
+					"Union `%v` should have only 1 union discriminator, but has 2: `%v` and `%v`",
```

Comment:
that's not critical if we can go go-style we would avoid warnings, so Go has convention not to capitalize error messages https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Reply:
Fixed in commit https://github.com/devfile/api/pull/142/commits/125c893a7a747127686c3ec73226f3d1eb02c533